### PR TITLE
support id, path or object of groups and project

### LINKED ISSUE #18: Ability to use project name as id
It seems the only way to limit things to a certain project is the int id.  And getting the int id is not easy on a large gitlab setup (took 2 minutes for me to `client.Projects.GetAsync()`)

The gitlab API does support escaped `group/project` endpoints.  I.e. `$"projects/{group}%2F{project}/repository/commits"`

I do not see a way to specify it in that fashion, but it would be very helpful

### LINKED ISSUE #33: Support Namespaced path encoding calls
GitLab API has two ways to access the project info (see [doc](https://docs.gitlab.com/ee/api/README.html#namespaced-path-encoding)).

1. By project ID (ex. https://gitlab.com/api/v4/projects/13083).
2. By namespace and project name (ex. https://gitlab.com/api/v4/projects/gitlab-org%2fgitlab-ce).

Add this functionality please.

### LINKED ISSUE #66: fixup request classes and use id/path as a object on method being called
I'd like to fix up the request classes to move id / path out of the request object and onto the parameter as type `object id`

So this pattern
https://github.com/nmklotas/GitLabApiClient/blob/2f5c1bcbf103e3ad7235af02ec2bc40c1177b718/src/GitLabApiClient/ProjectsClient.cs#L129-L133

is replaced with

```csharp
 public async Task<Project> UpdateAsync(object id, UpdateProjectRequest request) 
 { 
     Guard.NotNull(request, nameof(request)); 
     return await _httpFacade.Put<Project>($"projects/{id.ProjectIdOrPath}", request); 
 } 
```

```csharp
public static string ProjectIdOrPath(this object obj)
{
    switch (obj)
    {
        case null:
            throw new ArgumentException("ID or Path cannot be null");
        case int id:
            return id.ToString();
        case string path:
            return path.UrlEncode();
        case Project project:
        {
            int id = project.Id;
            if (id > 0)
            {
                return id.ToString();
            }

            string path = project.PathWithNamespace?.Trim();
            if (!string.IsNullOrEmpty(path))
            {
                return path.UrlEncode();
            }

            break;
        }
    }
    throw new ArgumentException($"Cannot determine project ID or Path from provided {obj.GetType().Name} instance. " +
                                "Must be int, string, Project instance");
}
```

reason being `id` is path parameter and not a body parameter. Same reason for all these query parameter PRs 😄

Would be great to switch all `int id`, `int projectId`, `int groupId` and `int userId`  out with `object id` so we can allow users to pass string without worrying about URL encoding or finding the damn project id.

This should help solve #18 and #33

### PR DESCRIPTION
**Pull request by casz**
Now the user of the library does not have to worry about path encoding of group and project.

I also tried to unify query parameters and requests objects and methods.

fixes #18 
fixes #33 
fixes #66

### CODE REVIEW

**Comment by casz** [comment]
@nmklotas Hi Julius.

I know this is a bigger PR, do you have any concerns with the changes. I know I am breaking the API a lot lately, I might look into cleaning up Milestone and Note API to be a separate API client with a interface since they are shared between Groups and Projects. See https://github.com/jdalrymple/node-gitlab/blob/master/src/core/templates/ResourceMilestones.ts.

Could also adopt more of the template resources like members.

I tried to align the methods.

**Comment by casz** [comment]
Hi @nmklotas I would appreciate a review or a merge 😅

Just ran into not being able to call
```
        public async Task<Project> GetProject(string projectId)
        {
            return await _client.Projects.GetAsync(projectId);
        }
```

**Comment by nmklotas** [comment]
Thanks for this big beutiful PR :) sorry for the delay.

I think we need to change few things.

1. Don't use projectBaseUrl() or other BaseUrl() methods, we should embrace the declarative nature of strings then it's very easy to mind map them to gitlab and its docs.

2. For parameters we should use a simple class, let's call it ProjectId, untyped object might cause problems.

it should have these properties:
```
ToString() overriden to create string for int, or encode if it's the path
implicit operator to create it from string
implicit operator to create it from int
```

so user can simply pass int or string and these operators would created class ProjectId object.
Then it's typed and easy to use and easy to maintain.

and since we are overriding ToString() it can be used in the path & it converts automatically.

what do you think?

**Comment by casz** [comment]
Sounds totally reasonable! 👍

**Comment by casz** [comment]
Thanks 👍

File: src/GitLabApiClient/Internal/Queries/CommitQueryBuilder.cs [hunk 1]
```diff
@@ -10,11 +11,27 @@ internal class CommitQueryBuilder : QueryBuilder<CommitQueryOptions>
...
+                Add("since", options.Since.Value);
+
+            if (options.Until.HasValue)
+                Add("until", options.Until.Value);
+
+            if (options.All.HasValue)
+                Add("all", options.All.Value);
+
+            if (options.WithStats.HasValue)
+                Add("all", options.WithStats.Value);
```
Comment (casz) [inline src/GitLabApiClient/Internal/Queries/CommitQueryBuilder.cs]: Copy paste oopsie!

```suggestion
                Add("with_stats", options.WithStats.Value);
```

File: src/GitLabApiClient/Internal/Queries/CommitQueryBuilder.cs [hunk 2]
```diff
@@ -10,11 +11,27 @@ internal class CommitQueryBuilder : QueryBuilder<CommitQueryOptions>
...
+
+            if (options.FirstParent.HasValue)
+                Add("all", options.FirstParent.Value);
```
Comment (casz) [inline src/GitLabApiClient/Internal/Queries/CommitQueryBuilder.cs]: Copy paste oopsie!

```suggestion
                Add("first_parent", options.FirstParent.Value);
```

**Comment by casz** [comment]
Travis CI come on gotta support dotnet core 3...

**Comment by casz** [comment]
There we go 😆 New and improved code coverage 👍

**Comment by casz** [comment]
Docker works fine locally, but Travis and appveyor both have blocking issues.

I think with appveyor using gitlab 11 which still uses postgres 10 we got closer :)

**Comment by casz** [comment]
Okay cleaned up the branch to only handle the API cleanup.

Moving the test setup fixes into a separate PR.

**Comment by casz** [comment]
@nmklotas Hope you don't mind Julius that I used my new found powers to release a new version 😄

**Pull request merged**